libcaf_net/test/net-test.cpp

```diff
@@ -118,6 +118,59 @@ ptrdiff_t mock_stream_transport::handle_input() {
   return result;
 }
 
+// -- mock_web_socket_app ------------------------------------------------------
+
+mock_web_socket_app::mock_web_socket_app(bool behave_as_server)
+  : behave_as_server(behave_as_server) {
+  // nop
+}
+
+caf::error mock_web_socket_app::start(caf::net::web_socket::lower_layer* ll) {
+  down = ll;
+  if (behave_as_server)
+    down->request_messages();
+  return caf::none;
+}
+
+caf::error
+mock_web_socket_app::accept(const caf::net::http::request_header& hdr) {
+  // Store the request information in cfg to evaluate them later.
+  auto& ws = cfg["web-socket"].as_dictionary();
+  put(ws, "method", to_rfc_string(hdr.method()));
+  put(ws, "path", std::string{hdr.path()});
+  put(ws, "query", hdr.query());
+  put(ws, "fragment", hdr.fragment());
+  put(ws, "http-version", hdr.version());
+  if (hdr.num_fields() > 0) {
+    auto& fields = ws["fields"].as_dictionary();
+    hdr.for_each_field([&fields](auto key, auto val) {
+      put(fields, std::string{key}, std::string{val});
+    });
+  }
+  return caf::none;
+}
+
+void mock_web_socket_app::abort(const caf::error& reason) {
+  if (expect_abort_) {
+    has_aborted = true;
+    err = reason;
+    expect_abort_ = false;
+    down->shutdown(reason);
+  } else {
+    CAF_FAIL("app::abort called: " << reason);
```

Comment:
"Hiding" a `CAF_FAIL` in a `.cpp` file outside of the test suite itself feels a bit off. I think dropping the `expect_abort_` logic here and just checking `has_aborted` in the tests will make them much easier to maintain.

Reply:
I left this as a safeguard since we might forget to expect that the connection did not abort, but I'll add it to the rest of the tests.